src/realm/query_engine.hpp

```diff
@@ -1744,27 +1744,49 @@ class StringNode : public StringNodeBase {
         TConditionFunction cond;
 
         for (size_t s = start; s < end; ++s) {
-            if constexpr (std::is_same_v<TConditionFunction, NotEqual>) {
+
+            // special handling for !=, <, <= , >, >= if the leaf is compressed and we have got a compressed string
+            // id.
+            if constexpr (realm::is_any_v<TConditionFunction, NotEqual, Greater, Less, GreaterEqual, LessEqual>) {
```

Comment:
All of these operators provide an integer overload for `bool operator()(int64_t v1, int64_t v2) const` which already does what you want, no need to reimplement it here. Just call it with the result of the comparison, and the other argument can be 0.